woocommerce/woocommerce-ios
# [Mobile Payments] Modal UI fixes

Fixes #4550 

This the requests from #4550. I've been fighting auto layout for a couple days already and was about to give up on some of the things, so I decided to put up a PR with the things I managed to fix. As I was writing up descriptions of the remaining issues I kept finding new solutions 😅 

## What's fixed

- Modal shown when connecting to reader has "Connecting to reader" in the title instead of "Scanning for reader"
- Modal shown when connecting to reader is displayed when manually connecting as well as when we auto connect
- Modals are always 280px wide and at least 382px high
- The top title on the modal uses the `body` font style
- The bottom title on the modal uses the `subheadline` font style
- Modals have 16px padding on all edges except the top one, which has 32px
- Adjusted spacing between bottom labels
- Adjusted spacing between buttons
- Adjusted layout so top title aligns to top, buttons align to bottom, and the rest of the content is evenly distributed between top labels, image, bottom labels, and buttons. Also, when there are no buttons, I added a spacer to the bottom, so that all the content is aligned to top 

Note that the spacing between the bottom labels in the "Insert Card" screen seems wrong. On Figma, I see a 2px space between labels, which is the spacing I set on Xcode, but the result is visually different. On Figma, both text boxes are 20px high and include some whitespace, on iOS the boxes are 18px and 15.67px respectively with the default font, so I could account for the difference and add an extra 6px of spacing, but I'm not sure if that would be what's expected with dynamic text.

Figma|Xcode|Runtime
-|-|-
![Screen Shot 2021-09-03 at 10 25 27](https://user-images.githubusercontent.com/8739/131974880-069338ec-2cf0-470d-b9af-9fb9ebae4911.png)|![Screen Shot 2021-09-03 at 10 25 53](https://user-images.githubusercontent.com/8739/131974954-d8a478ee-7df6-4237-a154-0bd7c5582e3a.png)|![Screen Shot 2021-09-03 at 10 30 42](https://user-images.githubusercontent.com/8739/131975599-a645f761-5281-4965-9471-bbed10500ec9.png)

## How it's fixed

I thought I'd leave a few notes, because reviewing diffs of XIB files is an exercise in decryption (even I'm having trouble remembering what I've changed exactly) 😩 

The behavior we want is that the top item (top labels) stick to the top, and the bottom item (buttons) sticks to the bottom. A stack view with a `fill` distribution won't do this by default and distribute the space evenly among all children, and these will typically be vertically centered in the available space. After a lot of trying, I adjusted the hugging priority to required (1000) in all the items I did not want to expand (top labels and their container, buttons and their container). For the bottom labels I see now that I left the priority at high (750), but I guess the behavior would be the same, as the rest of the items have a lower priority. The image and container for the bottom labels have low (250) priority, so they're the ones which will grow to fill the available space.

When there are no buttons at the bottom, it looks like the designs want the content gravitating towards the top, so I added a spacer view that's only visible when the buttons are not, and it has the lowest hugging priority (1). So if this spacer is visible, it will take any available extra space before any other view can grow.

I know I had to adjust something to make the buttons grow to the full width, but I can't find it in the diff now 😬 

Besides this, I added a bunch of Xcode labels for views, because it was impossible to understand the existing constraints otherwise (too many things named `StackView`)

## Screenshots

### Searching for reader

Screenshots | Designs
------------ | -------------
<img src="https://user-images.githubusercontent.com/8739/131972058-25ba220d-229d-45d5-a0f7-279057ff288c.png" width="375">|<img src="https://user-images.githubusercontent.com/28450084/124573583-fb001e80-de49-11eb-8393-22ab48c10b9e.png" width="375">

### Connect to reader

Screenshots | Designs
------------ | -------------
<img src="https://user-images.githubusercontent.com/8739/131972091-bca7ff40-e750-4554-a48d-6c50bd7f797d.png" width="375">|<img src="https://user-images.githubusercontent.com/28450084/124574020-59c59800-de4a-11eb-9a21-3be0029ea761.png" width="375">

### Connecting to reader

Screenshots | Designs
------------ | -------------
<img src="https://user-images.githubusercontent.com/8739/131972128-3bdae3d7-4de8-4cdc-94b9-3a99b78d2da2.png" width="375">|<img src="https://user-images.githubusercontent.com/8739/131972204-2290fc4d-f09e-4af1-a066-929c1f51d489.png" width="375">

### Insert card

Screenshots | Designs
------------ | -------------
<img src="https://user-images.githubusercontent.com/8739/131972231-b7416cee-511e-4f99-ba76-61d3b119a040.png" width="375">|<img src="https://user-images.githubusercontent.com/8739/131966470-5c171387-2886-49f8-a6da-545c06e49645.png" width="375">

### Remove card

Screenshots | Designs
------------ | -------------
<img src="https://user-images.githubusercontent.com/8739/131972276-310348ed-9f61-4ad5-9880-5a0ecc49a42a.png" width="375">|<img src="https://user-images.githubusercontent.com/8739/131973122-285fd3c5-8d36-41f1-b687-5221c05c8254.png" width="375">

### Payment completed

Screenshots | Designs
------------ | -------------
<img src="https://user-images.githubusercontent.com/8739/131974109-b6eb3bc6-c44f-42b3-b48e-0f7c27ef232b.png" width="375">|<img src="https://user-images.githubusercontent.com/8739/131972643-28b3ef3d-6bad-4d2e-a779-426712397356.png" width="375">

## To test

Using a store with payments available, and a reader that's not "known" (if necessary, go to Settings > In-Person Payments > Manage Card Reader and Connect then Disconnect manually):

1. Make sure your reader is turned off
2. Find an order that can be paid in person, tap on Collect payment
3. See **Scanning for reader** screen and make sure it matches the design
4. Turn on your reader
5. See **Connect to reader** screen and make sure it matches the design
6. Tap on Connect to reader
7. See **Connecting to reader** screen and make sure it matches the design
8. See **Insert Card** screen and make sure it matches the design
9. Tap your card to pay
10. See **Remove Card** screen and make sure it matches the design
11. See **Payment Completed** screen and make sure it matches the design

Update release notes:
